man/man3/Drawing.3
.\"
.\"    This program is free software; you can redistribute it and/or modify
.\"    it under the terms of the GNU General Public License as published by
.\"    the Free Software Foundation; either version 2 of the License, or
.\"    (at your option) any later version.
.\"
.\"    This program is distributed in the hope that it will be useful,
.\"    but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\"    GNU General Public License for more details.
.\"
.\"    You should have received a copy of the GNU General Public License
.\"    along with this program; if not, write to the Free Software
.\"    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.TH Drawing 3fe "7/11/94" "Version 2.21" "Finite Element Package"
.SH NAME
Drawing - general purpose drawing widget
.SH SYNOPSIS
.B "# include <X11/Intrinsic.h>"
.br
\fB# include "Drawing.h"\fR
.SH DESCRIPTION
These files define a general purpose drawing widget that is subclass of the
Athena \fISimple\fR widget.  The widget provides a scalable, two dimensional
drawing area with a cartesian coordinate system and an optional background
grid and pointer snap.  The widget is intended to provide mechanisms and not
policies.  For example, there is no "select rectangular area with mouse"
feature.  Instead, the application writer can specify callbacks associated
with pointer movement and mouse button actions.
.PP
The widget operations allow the creation, removal, and dynamic configuration
of figures.  The figures are maintained on a display list by the widget.
Figures are referenced by the application writer though the opaque type
\fIFigure\fR.  The current figure types include lines, polygons, rectangles,
arcs, text, pixmaps, bitmaps, and groups.  The properties associated with
figures include grouping, scaling, line width, line style, visibility, color,
and position.
.PP
.SH RESOURCES
The \fIDrawing\fR widget adds the following resources to the Athena
\fISimple\fR widget:
.TP 5
\fBXtNcoordinates\fR (class \fBXtCCoordinates\fR) XtRWidget
Specifies a \fIlabel\fR widget to be used in displaying the current
coordinates of the pointer within the widget or \s-1NULL\s+1 if no display is
desired.  Default value: \s-1NULL\s+1
.TP
\fBXtNgrid\fR (class \fBXtCGrid\fR) XtRBoolean
Specifies if the background grid should be displayed.  The grid consists of a
series of intersecting dashed lines lying behind all figures whose spacing is
specified by the \fBgridSize\fR resource.  Default value: \fBFalse\fR
.TP
\fBXtNgridSize\fR (class \fBXtCGridSize\fR) XtRFloat
Specifies the grid spacing in real units.  The grid lines will be drawn with
spacing according to the value of this resource if the \fBgrid\fR resource is
\fBTrue\fR.  Default value: 1.00
.TP
\fBXtNsnap\fR (class \fBXtCSnap\fR) XtRBoolean
Specifies if pointer snap should be enabled.  Pointer snap indicates that in
addition to reporting the actual coordinates of the pointer in a callback,
the coordinates lying closest to the pointer and along an invisible grid
whose spacing is given by the \fBsnapSize\fR resource should also be
reported.
.TP
\fBXtNSnapSize\fR (class \fBXtCSnapSize\fR) XtRFloat
Specifies the snap grid spacing in real units.  The invisible snap grid with
be spaced according to the value of this resource.  This resource has no
effect if the \fBsnap\fR resource is \fBFalse\fR.  Default value: 1.00
.TP
\fBXtNxMin\fR (class \fBXtCXMin\fR) XtRFloat
Specifies the minimum value for the x axis.  This value must be less than the
value specified by the \fBxMax\fR resource.  Default value: 0.0
.TP
\fBXtNxMax\fR (class \fBXtCXMax\fR) XtRFloat
Specifies the maximum value for the x axis.  This value must be greater than
the value specified by the \fBxMin\fR resource.  Default value: 10.0
.TP
\fBXtNyMin\fR (class \fBXtCYMin\fR) XtRFloat
Specifies the minimum value for the y axis.  This value must be less than the
value specified by the \fByMax\fR resource.  Default value: 0.0
.TP
\fBXtNyMax\fR (class \fBXtCYMax\fR) XtRFloat
Specifies the maximum value for the y axis.  This value must be greater than
the value specified by the \fByMin\fR resource.  Default value: 10.0
.TP
\fBXtNxScale\fR (class \fBXtCXScale\fR) XtRFloat
Specifies the scale for the x axis in pixels per real unit.  This value must
be greater than zero.  Default value: 50.0
.TP
\fBXtNyScale\fR (class \fBXtCYScale\fR) XtRFloat
Specifies the scale for the y axis in pixels per real unit.  This value must
be greater than zero.  Default value: 50.0
.TP
\fBXtNbuttonCallback\fR (class \fBXtCButtonCallback\fR) XtRCallback
A callback list whose procedures should be called upon a button action
(press, release, or drag).  These events include \fBButtonPress\fR,
\fBButtonRelease\fR, and \fBMotionNotify\fR events.  The call data to the
callback procedure will be a pointer to a \fBDrawingReport\fR structure which
includes both the snapped and unsnapped coordinates (the snapped coordinates
will be identical to the unsnapped coordinates if the \fBsnap\fR resource is
\fBFalse\fR) and a pointer to the corresponding \fBXEvent\fR structure.  The
topmost figure located under the pointer will also be included if the
\fBautoFind\fR resource is \fBTrue\fR and the corresponding event type is
\fBButtonPress\fR.  Default value: \s-1NULL\s+1
.TP
\fBXtNmotionCallback\fR (class \fBXtCMotionCallback\fR) XtRCallback
A callback list whose procedures should be called upon pointer motion in the
widget.  The call data to the callback procedure will be a pointer to a
\fBDrawingReport\fR structure as described above.  The corresponding event
type will always be \fBMotionNotify\fR.  Default value: \s-1NULL\s+1
.TP
\fBXtNinteractive\fR (class \fBXtCInteractive\fR) XtRBoolean
Specifies if interactive drawing mode is enabled.  In interactive mode, an
exclusive-or graphics context is used which provides for more efficient
redrawing and removal of figures than if the normal graphics context were
used.  However, the exclusive-or graphics context does not handle exposure
events properly.  Thus, interactive mode should be used for any interactive
operations such as moving a figure by following the pointer.  The general
policy should be to enable interactive mode, perform the necessary operation,
and then disable interactive mode.  The preferred way to set this resource is
through the \fBDW_SetInteractive\fR function.  Default value: \fBFalse\fR
.TP
\fBXtNautoFind\fR (class \fBXtCAutoFind\fR) XtRBoolean
Specifies if the display list is to be searched upon a button press.  If this
resource is \fBTrue\fR then the topmost figure located under the pointer, or
\s-1NULL\s+1 if no such figure exists, will be returned in the
\fBDrawingReport\fR structure.  The preferred way to set this resource is
through the \fBDW_SetAutoFind\fR function.  Default value: \fBFalse\fR
.TP
\fBXtNautoRedraw\fR (class \fBXtCAutoRedraw\fR) XtRBoolean
Normally the display list is automatically redrawn upon a change to the
display list though a \fBDW_RemoveFigure\fR, \fBDW_LowerFigure\fR, or
\fBDW_SetAttributes\fR operation.  If this resource is \fBFalse\fR then the
display list will not automatically be redrawn and the corresponding redraw
requests will be buffered.  Upon setting this resource to \fBTrue\fR, any
buffered requests will be processed.  The general policy should be to disable
redrawing before making substantial changes to the display list and to
re-enable redrawing once those changes are complete.  The preferred way to
set this resource is through the \fBDW_SetAutoRedraw\fR function.  Default
value: \fBTrue\fR
.PP
For convenience a function, \fBFloat2Arg\fR, is provided which converts a
floating point value to an \fBArg\fR suitable for use with \fBXtSetArg\fR.
.SH OPERATIONS
The \fIDrawing\fR widget provides a variety of functions for drawing figures,
altering the display list, changing the internal state of the widget, and
retrieving information about the widget:
.PP
.nf
.B "Figure DW_DrawLine (drawing, x1, y1, x2, y2)"
.fi
.RS
.PD 0
.TP 9
.B Widget
.B drawing;
.TP
.B float
.B x1, y1;
.TP
.B float
.B x2, y2;
.RE
.PD
.PP
\fBDW_DrawLine\fR draws a line in the current foreground color from
(\fIx1\fR,\fIy1\fR) to (\fIx2\fR,\fIy2\fR) using \fBXDrawLine\fR.  The figure

CHANGELOG
** Changes to FElt from 3.02 to 3.05 (first version released on SourceForge)

The following changes were made to the FElt syntax:
   1. changed the sense of coordinates for axisymmetric elements to
      make them easier to use with mesh generation.  x is still radial
      but now y is vertical/axial.  Previously z was vertical.  

The following bugs/things were fixed/changed:
   1. some small bugs in axisymmetric, ctg, htk, and iso_2d elements
      that I haven't been really good about keeping tack of.  Try
      the new version and see if your answers look better -- if they
      look worse then let me know ...
   2. you can now use the -details switch to get the coordinates of
      the points where stresses are calculated in the text display of
      results.
   3. added radial and axial directions to the load dialog in velvet
      and to the file writing routines.  They'd gotten left out when
      axisymmetric elements were originally implemented.
   4. finally resolved the colormap issue in velvet so you can 
      use color stress and displacement plots on > 8-bit displays.
      Take a look at src/Velvet/XCC.README to see how you too can
      accomplish this if you're X11 colormap impaired like us.
   5. dropped Geompk in favor of Triangle by J.R. Shewchuk.  Several
      reasons for the change: Geompk was always a little finicky,
      mostly because of our interface probably, but it was sometimes
      difficult to coax a mesh out of Geompk.  The interface to
      Triangle is very simple, it is coded in C which makes compilation
      and configuration on other architectures simple, and it seems
      to be pretty easy to get a decent mesh out of it.  Accordingly,
      the main input parameters have changed to target and alpha.
      target is the approximate number of desired elements in the mesh.
      alpha is used to generate a maximum bound on the area of a generated
      element: Amax of an elt = alpha*Atotal/target.  Triangle appears to
      have a lot of other options for mesh refinement, angle constraints
      etc., but none of these are made use of at this point.  Most of the
      hooks for Geompk are still available if someone really wants to
      switch back.  The call to GenerateTriMesh did not change for 
      applications, but the applications did change to account for
      the loss of Geompk parameters (min, max, angspc, angtol, etc.)
      and the addition of target and alpha.
     
** Changes to FElt from 3.00 or 3.01 to 3.02 

The following changes were made to the FElt syntax:
   1. hooks for a gravity vector in the analysis section.  Doesn't
      do anything yet, but the syntax is there ...
   2. "static-substitution" and "static-incremental" are now valid 
      analysis types
   3. a "load cases" section and accompanying syntax for named
      load cases
   4. syntax for new analysis parameters, "tolerance", "iterations",
      "relaxation", "load-steps".
   5. input-node= and input-dof= parameters in the analysis 
      parameters to allow for parametric studies in which you
      get a table (or plot) of output that explores displacements
      as a function of force level at a single input point
   6. axial and radial are now valid directions for distributed loads

The following bugs/things were fixed/changed:
   1. the transform matrix for beam3d elements was incorrect if elements
      had nodes which were out of the x-y plane.
   2. several problems with the new Makefile scheme were corrected.
   3. velvet was probably getting the aspect ratios wrong in drawing 
      3d structures due to an error in the min/max checking.
   4. the spectrum components of force structures were not being
      initialized to either zero or NULL
   5. save's in velvet were actually doing dumps (save all's), minor, 
      but potentially annoying.
   6. there was still a problem with the lumped mass matrices for 
      htk elements.
   7. the new PrintGlobalMatrices routine took a file pointer but
      didn't actually use it; things were just going to stdout.
   8. internal error in burlap's interface wit the matrix lib

lib/Widgets/Drawing.c
/*
    This file is part of the FElt finite element analysis package.
    Copyright (C) 1993-2000 Jason I. Gobat and Darren C. Atkinson

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/

/************************************************************************
 * File:	Drawing.c						*
 *									*
 * Description:	This file contains the widget functions for the Drawing	*
 *		widget.							*
 ************************************************************************/

# include <stdio.h>
# include <math.h>
# include <X11/IntrinsicP.h>
# include <X11/StringDefs.h>
# include <X11/Xaw/XawInit.h>
# include "DrawingP.h"


/* Private functions */

static void    ClassInitialize (void);
static void    Initialize (Widget request, Widget new, ArgList argv, Cardinal *argc);
static void    Realize (Widget gw, XtValueMask *valuemaskp, XSetWindowAttributes *attr);
static void    Redisplay (Widget gw, XEvent *event, Region region);
static void    Destroy (Widget gw);
static Boolean SetValues (Widget old, Widget request, Widget new, ArgList argv, Cardinal *argc);


/* Resource defaults */

# define DefaultGrid		FALSE
# define DefaultGridSize	1.00
# define DefaultSnap		FALSE
# define DefaultSnapSize	0.25
# define DefaultXMin		0.00
# define DefaultXMax		10.0
# define DefaultYMin		0.00
# define DefaultYMax		10.0
# define DefaultXScale		50.0
# define DefaultYScale		50.0
# define DefaultCursor		"crosshair"


/* Private variables */

static float defaultgridsize	= DefaultGridSize;
static float defaultsnapsize	= DefaultSnapSize;
static float defaultxmin	= DefaultXMin;
static float defaultxmax	= DefaultXMax;
static float defaultymin	= DefaultYMin;
static float defaultymax	= DefaultYMax;
static float defaultxscale	= DefaultXScale;
static float defaultyscale	= DefaultYScale;


/* Resources */

# define offset(field) XtOffsetOf(DrawingRec, field)

static XtResource resources[] = {
{XtNforeground, XtCForeground, XtRPixel, sizeof (Pixel),
    offset(drawing.foreground), XtRString, (XtPointer) XtDefaultForeground},
{XtNfont, XtCFont, XtRFontStruct, sizeof (XFontStruct *),
    offset(drawing.font), XtRString, (XtPointer) XtDefaultFont},
{XtNcoordinates, XtCCoordinates, XtRWidget, sizeof (Widget),
    offset(drawing.coord), XtRWidget, (XtPointer) NULL},
{XtNgrid, XtCGrid, XtRBoolean, sizeof (Boolean),
    offset(drawing.grid), XtRImmediate, (XtPointer) DefaultGrid},
{XtNgridSize, XtCGridSize, XtRFloat, sizeof (float),
    offset(drawing.gridSize), XtRFloat, (XtPointer) &defaultgridsize},
{XtNsnap, XtCSnap, XtRBoolean, sizeof (Boolean),
    offset(drawing.snap), XtRImmediate, (XtPointer) DefaultSnap},
{XtNsnapSize, XtCSnapSize, XtRFloat, sizeof (float),
    offset(drawing.snapSize), XtRFloat, (XtPointer) &defaultsnapsize},
{XtNxMin, XtCXMin, XtRFloat, sizeof (float),
    offset(drawing.xMin), XtRFloat, (XtPointer) &defaultxmin},
{XtNxMax, XtCXMax, XtRFloat, sizeof (float),
    offset(drawing.xMax), XtRFloat, (XtPointer) &defaultxmax},
{XtNyMin, XtCYMin, XtRFloat, sizeof (float),
    offset(drawing.yMin), XtRFloat, (XtPointer) &defaultymin},
{XtNyMax, XtCYMax, XtRFloat, sizeof (float),
    offset(drawing.yMax), XtRFloat, (XtPointer) &defaultymax},
{XtNxScale, XtCXScale, XtRFloat, sizeof (float),
    offset(drawing.xScale), XtRFloat, (XtPointer) &defaultxscale},
{XtNyScale, XtCYScale, XtRFloat, sizeof (float),
    offset(drawing.yScale), XtRFloat, (XtPointer) &defaultyscale},
{XtNbuttonCallback, XtCButtonCallback, XtRCallback, sizeof (XtPointer),
    offset(drawing.button), XtRImmediate, (XtPointer) NULL},
{XtNmotionCallback, XtCMotionCallback, XtRCallback, sizeof (XtPointer),
    offset(drawing.motion), XtRImmediate, (XtPointer) NULL},
{XtNinteractive, XtCInteractive, XtRBoolean, sizeof (Boolean),
    offset(drawing.interactive), XtRImmediate, (XtPointer) FALSE},
{XtNautoFind, XtCAutoFind, XtRBoolean, sizeof (Boolean),
    offset(drawing.search), XtRImmediate, (XtPointer) FALSE},
{XtNautoRedraw, XtCAutoRedraw, XtRBoolean, sizeof (Boolean),
    offset(drawing.redraw), XtRImmediate, (XtPointer) TRUE},
{XtNcursor, XtCCursor, XtRCursor, sizeof (Cursor),
    offset(simple.cursor), XtRString, (XtPointer) DefaultCursor},
};
# undef offset


DrawingClassRec drawingClassRec = {
  { /* core fields */
    /* superclass		*/	(WidgetClass) &simpleClassRec,
    /* class_name		*/	"Drawing",
    /* widget_size		*/	sizeof (DrawingRec),
    /* class_initialize		*/	ClassInitialize,
    /* class_part_initialize	*/	NULL,
    /* class_inited		*/	FALSE,
    /* initialize		*/	Initialize,
    /* initialize_hook		*/	NULL,
    /* realize			*/	Realize,
    /* actions			*/	NULL,
    /* num_actions		*/	0,
    /* resources		*/	resources,
    /* num_resources		*/	XtNumber (resources),
    /* xrm_class		*/	NULLQUARK,
    /* compress_motion		*/	FALSE,
    /* compress_exposure	*/	TRUE,
    /* compress_enterleave	*/	TRUE,
    /* visible_interest		*/	TRUE,
    /* destroy			*/	Destroy,
    /* resize			*/	NULL,
    /* expose			*/	Redisplay,
    /* set_values		*/	SetValues,
    /* set_values_hook		*/	NULL,
    /* set_values_almost	*/	XtInheritSetValuesAlmost,
    /* get_values_hook		*/	NULL,
    /* accept_focus		*/	NULL,
    /* version			*/	XtVersion,
    /* callback_private		*/	NULL,
    /* tm_table			*/	NULL,
    /* query_geometry		*/	XtInheritQueryGeometry,
    /* display_accelerator	*/	XtInheritDisplayAccelerator,
    /* extension		*/	NULL
  },
  { /* simple fields */
    /* change_sensitive		*/	XtInheritChangeSensitive
  },
  { /* drawing fields */
    /* empty			*/	0
  }
};


WidgetClass drawingWidgetClass = (WidgetClass) &drawingClassRec;


/* Private functions */

/************************************************************************
   Function:	SnapCoord
   Description:	Snap the coordinates to the snap grid if enabled.
 ************************************************************************/

static void SnapCoord (DrawingWidget dw, float *x, float *y)
{
    double snap;

    if (dw -> drawing.snap == True) {
	snap = dw -> drawing.snapSize;
	*x = (int) ((*x + snap / (*x >= 0 ? 2 : -2)) / snap) * snap;
	*y = (int) ((*y + snap / (*y >= 0 ? 2 : -2)) / snap) * snap;
    }
}



/************************************************************************
   Function:	MotionHandler
   Description:	Handles motion events when a button is not depressed.
 ************************************************************************/

static void MotionHandler (Widget gw, XtPointer clientData, XEvent *event, Boolean *cont)
{
    DrawingReport report;
    DrawingWidget dw;
    Arg		  al [1];
    char	  string [40];
    float	  realx;
    float	  realy;
    int		  x;
    int		  y;


    dw = (DrawingWidget) gw;
    x = event -> xmotion.x;
    y = event -> xmotion.y;

    if (x < 0 || x > (int) dw -> drawing.width)
	return;

    if (y < 0 || y > (int) dw -> drawing.height)
	return;

    realx = RealX (x);
    realy = RealY (y);

    report.event       = event;
    report.unsnapped.x = realx;
    report.unsnapped.y = realy;
    report.snapped.x   = realx;
    report.snapped.y   = realy;
    SnapCoord (dw, &report.snapped.x, &report.snapped.y);

    if (dw -> drawing.coord != NULL) {
	sprintf (string, "%.*f,%.*f", dw -> drawing.xprecision,
	report.snapped.x, dw -> drawing.yprecision, report.snapped.y);
	XtSetArg (al [0], XtNlabel, string);
	XtSetValues (dw -> drawing.coord, al, 1);
    }

    XtCallCallbacks (gw, XtNmotionCallback, (XtPointer) &report);
}


/************************************************************************
   Function:	ButtonHandler
   Description:	Handles mouse events when a button is depressed.
 ************************************************************************/

static void ButtonHandler (Widget gw, XtPointer clientData, XEvent *event, Boolean *cont)
{
    static DrawingReport report;
    DrawingWidget	 dw;
    float		 realx;
    float		 realy;
    int			 x;
    int			 y;


    dw = (DrawingWidget) gw;
    x = event -> xmotion.x;
    y = event -> xmotion.y;

    if (x < 0)
	x = 0;
    else if (x > (int) dw -> drawing.width)
	x = dw -> drawing.width;

    if (y < 0)
	y = 0;
    else if (y > (int) dw -> drawing.height)
	y = dw -> drawing.height;


    realx = RealX (x);
    realy = RealY (y);

    report.event       = event;
    report.unsnapped.x = realx;
    report.unsnapped.y = realy;
    report.snapped.x   = realx;
    report.snapped.y   = realy;
    SnapCoord (dw, &report.snapped.x, &report.snapped.y);


    if (event -> type == ButtonPress) {
         if (dw -> drawing.search == True)
              report.figure = DW_FindFigure (gw, realx, realy);
         else
              report.figure = NULL;
    }

    XtCallCallbacks (gw, XtNbuttonCallback, (XtPointer) &report);
}


/************************************************************************
   Function:	SetClipRegion
   Description:	Sets the clip region to the size of the widget.
 ************************************************************************/

static void SetClipRegion (DrawingWidget dw)
{
    Region     nullRegion;
    XRectangle rect;


    rect.x = 0;
    rect.y = 0;
    rect.width = dw -> drawing.width;
    rect.height = dw -> drawing.height;

    nullRegion = dw -> drawing.nullRegion;
    XIntersectRegion (nullRegion, dw -> drawing.region, dw -> drawing.region);
    XUnionRectWithRegion (&rect, dw -> drawing.region, dw -> drawing.region);
}


/************************************************************************
   Function:	DrawGrid
   Description:	Draw the grid on the window.
 ************************************************************************/

static void DrawGrid (Widget gw)
{
    int		  coord;
    float	  size;
    float	  x;
    float	  y;
    float	  min;
    float	  max;
    float	  scale;
    Display	 *display;
    DrawingWidget dw;


    dw = (DrawingWidget) gw;
    display = XtDisplay (gw);

    min   = dw -> drawing.xMin;
    max   = dw -> drawing.xMax;
    size  = dw -> drawing.gridSize;
    scale = dw -> drawing.xScale;

    for (x = min - size; x <= max + size; x += size) {
	coord = ((int) (x / size) * size - min) * scale;
	XDrawLine (display, XtWindow (gw), dw -> drawing.gridgc, coord, 0,
		   coord, dw -> drawing.height);
    }

    min   = dw -> drawing.yMin;
    max   = dw -> drawing.yMax;
    scale = dw -> drawing.yScale;

    for (y = min - size; y <= max + size; y += size) {
	coord = (max - (int) (y / size) * size) * scale;
	XDrawLine (display, XtWindow (gw), dw -> drawing.gridgc, 0, coord,
		   dw -> drawing.width, coord);
    }
}


/************************************************************************
   Function:	DrawList
   Description:	Draws the display list.
 ************************************************************************/

static void DrawList (DrawingWidget dw)
{
    Figure fig;


    for (fig = dw -> drawing.head; fig != NULL; fig = fig -> next)
	DW_DrawFigure (dw, fig);
}


/************************************************************************
   Function:	ScaleList
   Description:	Scales the display list.
 ************************************************************************/

static void ScaleList (DrawingWidget dw)
{
    Figure fig;


    for (fig = dw -> drawing.head; fig != NULL; fig = fig -> next)
	DW_ScaleFigure (dw, fig);
}


/* Widget class functions */

/************************************************************************
   Function:	ClassInitialize
   Description: Initializes the widget class.
 ************************************************************************/

static void ClassInitialize (void)
{
    XawInitializeWidgetSet ( );
}


/************************************************************************
   Function:	Initialize
   Description:	Initializes the widget.
 ************************************************************************/

static void Initialize (Widget request, Widget new, ArgList argv, Cardinal *argc)
{
    Display	 *display;
    XtArgVal	  value;
    CacheData	  data;
    DrawingWidget dw;


    dw = (DrawingWidget) new;
    display = XtDisplay (new);


    if (dw -> drawing.xMin >= dw -> drawing.xMax) {
	dw -> drawing.xMin = DefaultXMin;
	dw -> drawing.xMax = DefaultXMax;
    }

    if (dw -> drawing.yMin >= dw -> drawing.yMax) {
	dw -> drawing.yMin = DefaultYMin;
	dw -> drawing.yMax = DefaultYMax;
    }

    if (dw -> drawing.xScale <= 0)
	dw -> drawing.xScale = DefaultXScale;

    if (dw -> drawing.yScale <= 0)
	dw -> drawing.yScale = DefaultYScale;

    if (dw -> drawing.gridSize <= 0)
	dw -> drawing.gridSize = DefaultGridSize;

    if (dw -> drawing.snapSize <= 0)
	dw -> drawing.snapSize = DefaultSnapSize;


    dw -> drawing.width  = (dw -> drawing.xMax - dw -> drawing.xMin) *
			    dw -> drawing.xScale;
    dw -> drawing.height = (dw -> drawing.yMax - dw -> drawing.yMin) *
			    dw -> drawing.yScale;

    dw -> core.width  = dw -> drawing.width;
    dw -> core.height = dw -> drawing.height;

    dw -> drawing.gridgc = NULL;
    dw -> drawing.drawgc = NULL;
    dw -> drawing.head   = NULL;
    dw -> drawing.tail   = NULL;

    dw -> drawing.xprecision  = ceil (log10 (dw -> drawing.xScale));
    dw -> drawing.yprecision  = ceil (log10 (dw -> drawing.yScale));
    dw -> drawing.line_width  = 0;
    dw -> drawing.line_style  = DW_LineSolid;
    dw -> drawing.search      = False;
    dw -> drawing.interactive = False;
    dw -> drawing.redraw      = True;
    dw -> drawing.last_style  = -1;
    dw -> drawing.last_fg     = -1;
    dw -> drawing.last_width  = 0;
    dw -> drawing.last_font   = 0;

    dw -> drawing.color_cache = DW_CacheCreate ( );
    dw -> drawing.font_cache = DW_CacheCreate ( );

    value = (XtArgVal) dw -> drawing.foreground;
    data = DW_CacheInsert (dw -> drawing.color_cache, "initial", value);
    dw -> drawing.fg = dw -> drawing.foreground;
    dw -> drawing.color_data = data;

    value = (XtArgVal) dw -> drawing.font;
    data = DW_CacheInsert (dw -> drawing.font_cache, "initial", value);
    dw -> drawing.font_data = data;
}


/************************************************************************
   Function:	Realize
   Description: Realizes the widget by realizing the superclass.  Adds
		events handlers and creates the GCs.
 ************************************************************************/

static void Realize (Widget gw, XtValueMask *valuemaskp, XSetWindowAttributes *attr)
{
    Display	 *display;
    DrawingWidget dw;
    unsigned long mask;
    XGCValues	  values;
    static char   dashes [ ] = {1, 4};


    dw = (DrawingWidget) gw;
    display = XtDisplay (gw);

    (*drawingWidgetClass -> core_class.superclass -> core_class.realize)
	(gw, valuemaskp, attr);

    mask = ButtonPressMask | ButtonReleaseMask | OwnerGrabButtonMask;
    mask |= Button1MotionMask | Button2MotionMask | Button3MotionMask;
    XtAddEventHandler (gw, mask, False, ButtonHandler, NULL);
    XtAddEventHandler (gw, PointerMotionMask, False, MotionHandler, NULL);


    mask = GCFunction | GCForeground | GCBackground | GCLineStyle;

    values.function   = GXcopy;
    values.line_style = LineOnOffDash;
    values.background = dw -> core.background_pixel;
    values.foreground = dw -> drawing.foreground;

    dw -> drawing.gridgc = XCreateGC (display, XtWindow (gw), mask, &values);
    XSetDashes (display, dw -> drawing.gridgc, 0, dashes, 2);


    mask = GCFunction | GCForeground | GCBackground;

    values.function   = GXcopy;
    values.background = dw -> core.background_pixel;
    values.foreground = dw -> drawing.foreground;

    dw -> drawing.drawgc = XCreateGC (display, XtWindow (gw), mask, &values);


    mask = GCFunction | GCForeground | GCBackground;

    values.function   = GXxor;
    values.background = dw -> core.background_pixel;
    values.foreground = dw -> drawing.foreground ^ values.background;

    if (values.foreground == values.background)
	values.function = GXequiv;

    dw -> drawing.intergc = XCreateGC (display, XtWindow (gw), mask, &values);


    dw -> drawing.region = XCreateRegion ( );
    dw -> drawing.nullRegion = XCreateRegion ( );
    dw -> drawing.bufferRegion = XCreateRegion ( );
    SetClipRegion (dw);
}


/************************************************************************
   Function:	Destroy
   Description:	Destroys private resources.
 ************************************************************************/

static void Destroy (Widget gw)
{
    Figure	  fig;
    Figure	  next;
    Display	 *display;
    DrawingWidget dw;


    dw = (DrawingWidget) gw;
    display = XtDisplay (gw);

    if (dw -> drawing.region)
	XDestroyRegion (dw -> drawing.region);

    if (dw -> drawing.gridgc)
	XFreeGC (display, dw -> drawing.gridgc);

    if (dw -> drawing.drawgc)
	XFreeGC (display, dw -> drawing.drawgc);

    if (dw -> drawing.intergc)
	XFreeGC (display, dw -> drawing.intergc);

    fig = dw -> drawing.head;
    while (fig != NULL) {
	next = fig -> next;
	DW_DestroyFigure (fig);
	fig = next;
    }

    DW_CacheDestroy (dw -> drawing.font_cache);
    DW_CacheDestroy (dw -> drawing.color_cache);
}


/************************************************************************
   Function:	Redisplay
   Description:	Redisplays the widget.
 ************************************************************************/

static void Redisplay (Widget gw, XEvent *event, Region region)
{
    Display	 *display;
    DrawingWidget dw;


    dw = (DrawingWidget) gw;
    display = XtDisplay (gw);

    if (dw -> core.visible == True) {
	XIntersectRegion (region, dw -> drawing.region, region);
	XSetRegion (display, dw -> drawing.gridgc, region);
	XSetRegion (display, dw -> drawing.drawgc, region);

	if (dw -> drawing.grid == True)
	    DrawGrid (gw);

	DrawList (dw);
	XSetRegion (display, dw -> drawing.gridgc, dw -> drawing.region);
	XSetRegion (display, dw -> drawing.drawgc, dw -> drawing.region);
    }
}


/************************************************************************
   Function:	SetValues
   Description:	Checks validity of new values and updates private
		information.
 ************************************************************************/

static Boolean SetValues (Widget old, Widget request, Widget new, ArgList argv, Cardinal *argc)
{
    Boolean       resize;
    Boolean	  rescale;
    Boolean       redisplay;
    DrawingWidget odw;
    DrawingWidget ndw;


    odw = (DrawingWidget) old;
    ndw = (DrawingWidget) new;

    resize    = False;
    rescale   = False;
    redisplay = False;


    if (ndw -> drawing.xMin >= ndw -> drawing.xMax) {
	ndw -> drawing.xMin = odw -> drawing.xMin;
	ndw -> drawing.xMax = odw -> drawing.xMax;
    }

    if (ndw -> drawing.yMin >= ndw -> drawing.yMax) {
	ndw -> drawing.yMin = odw -> drawing.yMin;
	ndw -> drawing.yMax = odw -> drawing.yMax;
    }

    if (ndw -> drawing.xScale <= 0)
	ndw -> drawing.xScale = odw -> drawing.xScale;

    if (ndw -> drawing.yScale <= 0)
	ndw -> drawing.yScale = odw -> drawing.yScale;

    if (ndw -> drawing.gridSize <= 0)
	ndw -> drawing.gridSize = odw -> drawing.gridSize;

    if (ndw -> drawing.snapSize <= 0)
	ndw -> drawing.snapSize = odw -> drawing.snapSize;


    ndw -> drawing.width  = (ndw -> drawing.xMax - ndw -> drawing.xMin) *
			     ndw -> drawing.xScale;
    ndw -> drawing.height = (ndw -> drawing.yMax - ndw -> drawing.yMin) *
			     ndw -> drawing.yScale;

    ndw -> drawing.xprecision  = ceil (log10 (ndw -> drawing.xScale));
    ndw -> drawing.yprecision  = ceil (log10 (ndw -> drawing.yScale));


    if (odw -> drawing.xMin != ndw -> drawing.xMin)
	redisplay = rescale = True;
    if (odw -> drawing.xMax != ndw -> drawing.xMax)
	redisplay = rescale = True;
    if (odw -> drawing.yMin != ndw -> drawing.yMin)
	redisplay = rescale = True;
    if (odw -> drawing.yMax != ndw -> drawing.yMax)
	redisplay = rescale = True;
    if (odw -> drawing.xScale != ndw -> drawing.xScale)
	redisplay = rescale = True;
    if (odw -> drawing.yScale != ndw -> drawing.yScale)
	redisplay = rescale = True;
    if (odw -> drawing.width != ndw -> drawing.width)
	redisplay = resize = True;
    if (odw -> drawing.height != ndw -> drawing.height)
	redisplay = resize = True;
    if (odw -> drawing.gridSize != ndw -> drawing.gridSize)
	redisplay = resize = True;
    if (odw -> drawing.grid != ndw -> drawing.grid)
	redisplay = True;

    if (odw -> drawing.interactive != ndw -> drawing.interactive)
	DW_SetInteractive (new, ndw -> drawing.interactive);

    if (odw -> drawing.redraw != ndw -> drawing.redraw)
	DW_SetAutoRedraw (new, ndw -> drawing.redraw);


    ndw -> core.width = ndw -> drawing.width;
    ndw -> core.height = ndw -> drawing.height;


    if (resize == True)
	SetClipRegion (ndw);

    if (rescale == True)
	ScaleList (ndw);

    return redisplay;
}
